src/core/net/srp_server.cpp

```diff
@@ -1526,20 +1541,56 @@ Server::Service *Server::Host::AddService(const char *aFullName)
     return service;
 }
 
-void Server::Host::RemoveAndFreeService(Service *aService)
+void Server::Host::RemoveService(Service *aService, bool aRetainName, bool aNotifyServiceHandler)
 {
-    if (aService != nullptr)
+    const Server &server = Get<Server>();
+
+    VerifyOrExit(aService != nullptr);
+
+    aService->mIsDeleted = true;
+
+    if (aRetainName)
+    {
+        aService->ClearResources();
+        otLogInfoSrp("[server] remove service '%s' (but retain its name)", aService->mFullName);
+    }
+    else
+    {
+        otLogInfoSrp("[server] fully remove service '%s'", aService->mFullName);
+    }
+
+    if (aNotifyServiceHandler && server.mServiceUpdateHandler != nullptr)
+    {
+        LinkedList<Service> remainingServices = mServices;
+
+        mServices.Clear();
+        IgnoreError(mServices.Add(*aService));
```

Comment:
@wgtdkp, I just noticed this. I think this may cause some issue since `aService`  (which is already in the `remainingServices` linked list) is being placed into a new list and this can mess up the original `remainingServcies` list. 

One solution can be that we remove it from `remainingServices` before adding it here to `mServcies` and then re-add it to `reminingServices`.

Reply:
Good catch! I revised accordingly, please re-review.